yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,412 @@
+                encoder_mask (torch.Tensor): mask for the batch of encoded
+                        input symbols. Shape: (batch_size x src_len).
+
+        Returns:
+                predictions (torch.Tensor): tensor of predictions of shape
+                        (batch_size x pred_seq_len).
+                likelihood (torch.Tensor): tensor of final likelihoods per
+                    prediction step. Shape (batch_size x 1 x src_len).
+        """
+        batch_size, _ = encoder_mask.shape
+        log_probs, alignment_matrix, decoder_hiddens = self.init_decoding(
+            encoder_out, encoder_mask
+        )
+        predictions, likelihood = self.greedy_step(
+            log_probs, alignment_matrix[:, 0].unsqueeze(1)
+        )
+        finished = (
+            torch.zeros(batch_size, device=self.device).bool().unsqueeze(-1)
+        )
+        for t in range(self.max_target_length):
+            symbol = predictions[:, t]
+            log_probs, alignment_matrix, decoder_hiddens = self.decode_step(
+                symbol.unsqueeze(-1),
+                decoder_hiddens,
+                encoder_out,
+                encoder_mask,
+            )
+            likelihood = likelihood + alignment_matrix.transpose(1, 2)
+            likelihood = likelihood.logsumexp(dim=-1, keepdim=True).transpose(
+                1, 2
+            )
+            pred, likelihood = self.greedy_step(log_probs, likelihood)
+            finished = finished | (pred == self.end_idx)
+            if finished.all().item():
+                break
+            # If finished decoding, pads.
+            pred = torch.where(
+                ~finished, pred, torch.tensor(self.end_idx, device=self.device)
+            )
+            predictions = torch.cat((predictions, pred), dim=-1)
+            # Updates likelihood emissions.
+            likelihood = likelihood + self._gather_at_idx(log_probs, pred)
+        return predictions, likelihood
+
+    def greedy_step(self, log_probs, likelihood):
+        # Updates current likelihood with greedy probability prediction.
+        tgt_prob = likelihood + log_probs.transpose(1, 2)
+        tgt_prob = tgt_prob.logsumexp(dim=-1)
+        tgt = torch.max(tgt_prob, dim=-1)[1]
+        return tgt.unsqueeze(-1), likelihood
+
+    def _gather_at_idx(self, prob, tgt):
+        # Collects probability of tgt index across all states in prob.
+        batch_size, src_seq_len, vocab_size = prob.shape
+        idx = tgt.view(-1, 1).expand(batch_size, src_seq_len).unsqueeze(-1)
+        output = torch.gather(prob, -1, idx).view(batch_size, 1, src_seq_len)
+        idx = idx.view(batch_size, 1, src_seq_len)
+        pad_mask = (idx != self.pad_idx).float()
+        return output * pad_mask
+
+    def _apply_mono_mask(self, alignment):
+        # Applies mask to enforce monotonic alignment.
+        align_mask = torch.ones_like(alignment[0]).triu().unsqueeze(0)
+        align_mask = (align_mask - 1) * -numpy.log(1e-7)
```

Comment:
For my own understanding, can you explain why we want 0, -1 here rather than 0, 1? Otherwise I understand whats going on and it is really really nicely implemented :).

Reply:
Lemme revisit this, I was cribbing from Shinje and Ryan.

Reply:
Not directly related but maybe you want to do this at the top:

```python
# Note I am using math here because there's no reason to use numpy.
ALIGNMENT_EPSILON = -math.log(1e-7)  
```

and then use that as a constant later.

Reply:
@Adamits aight so `triu` sets bottom triangular to 0, and `ones_like` would need to be multiplied by a negative factor to allow 0's and 1's. It would work just the same but I want to say that multiplying the entire matrix by negative one instead of only half is more expensive an operation (I imagine backend pytorch short circuits 0 multiplication). That's only reason I can imagine.

What annoys me is that you need to do the numpy.log thing. When I try a one line `tensor.where(tensor, -float("inf")), it messes up gradient calculations. Any thoughts on that?